ccl/sqlproxyccl: TestDirectoryConnect failed

ccl/sqlproxyccl.TestDirectoryConnect [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/10649859?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/10649859?buildTab=artifacts#/) on release-23.1 @ [624d9dea7f1296af60d16bbb45cc9d1259b3a1be](https://github.com/cockroachdb/cockroach/commits/624d9dea7f1296af60d16bbb45cc9d1259b3a1be):

Fatal error:

```
panic: test timed out after 4m55s

```
Stack: 

```
goroutine 42056 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x8e
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x32
```
<details><summary>Log preceding fatal error</summary>
<p>

```
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:314 +0xef
* 
* goroutine 34864 [chan receive, 2 minutes]:
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x7439140, 0xc00ac98510})
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:324 +0x9f
* runtime/pprof.Do({0x74390d0?, 0xc00022a000?}, {{0xc000166a00?, 0xc0050fb180?, 0xc0050fb1d0?}}, 0xc00207c7a8)
* 	GOROOT/src/runtime/pprof/runtime.go:40 +0xa3
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0x0?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:322 +0x58
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:314 +0xef
* 
* goroutine 34746 [select]:
* github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue.func2()
* 	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:388 +0x86
* created by github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue
* 	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:385 +0x33f
* 
* goroutine 34741 [select, 2 minutes]:
* github.com/cockroachdb/cockroach/pkg/util/log.(*fileSink).gcDaemon(0xc00772b2b0, {0x7439098, 0xc001cd79c0})
* 	github.com/cockroachdb/cockroach/pkg/util/log/file_log_gc.go:25 +0x7f
* created by github.com/cockroachdb/cockroach/pkg/util/log.ApplyConfig
* 	github.com/cockroachdb/cockroach/pkg/util/log/flags.go:306 +0x10aa
* 
* goroutine 34737 [sleep]:
* time.Sleep(0x4c4b40)
* 	GOROOT/src/runtime/time.go:195 +0x135
* github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Quiesce(0xc00790b080, {0x74390d0?, 0xc00022a000})
* 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:584 +0x1a5
* github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc00790b080, {0x74390d0, 0xc00022a000})
* 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:526 +0x11d
* github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.TestDirectoryConnect(0xc006f49380)
* 	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/proxy_handler_test.go:1447 +0xc8f
* testing.tRunner(0xc006f49380, 0x6204ef8)
* 	GOROOT/src/testing/testing.go:1446 +0x10b
* created by testing.(*T).Run
* 	GOROOT/src/testing/testing.go:1493 +0x35f
* 
* goroutine 34794 [chan receive, 2 minutes]:
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x7439140, 0xc008f172f0})
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:324 +0x9f
* runtime/pprof.Do({0x74390d0?, 0xc00022a000?}, {{0xc000166a00?, 0xc003d2d810?, 0xc006d235b0?}}, 0xc0015c97a8)
* 	GOROOT/src/runtime/pprof/runtime.go:40 +0xa3
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0xc0084aa720?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:322 +0x58
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:314 +0xef
* 
*
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDirectoryConnect.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-29014

--- a/pkg/ccl/sqlproxyccl/proxy_handler_test.go
+++ b/pkg/ccl/sqlproxyccl/proxy_handler_test.go
@@ -1389,13 +1389,17 @@ func TestDirectoryConnect(t *testing.T) {
 		})
 	})
 
+	// Drain the tenant server gracefully. This is a workaround for #106537.
+	// Draining the server allows the server to delete the sql instance row.
+	require.NoError(t, tenants[0].DrainClients(ctx))
+
 	// Stop the directory server and the tenant SQL process started earlier.
 	// This tests whether the proxy can recover when the directory server and
 	// SQL pod restarts.
 	tds.Stop(ctx)
 	tenantStopper.Stop(ctx)
 
-	// Drain old pod and add a new one before starting the directory server.
+	// Drain the old pod and add a new one before starting the directory server.
 	tds.DrainPod(tenantID, tenants[0].SQLAddr())
 	tenants = startTestTenantPods(ctx, t, s, tenantID, 1, base.TestingKnobs{})
 	tds.AddPod(tenantID, &tenant.Pod{
@@ -1407,15 +1411,14 @@ func TestDirectoryConnect(t *testing.T) {
 	require.NoError(t, tds.Start(ctx))
 
 	t.Run("successful connection after restart", func(t *testing.T) {
-		require.Eventually(t, func() bool {
+		testutils.SucceedsSoon(t, func() error {
 			conn, err := pgx.Connect(ctx, connectionString)
 			if err != nil {
-				return false
+				return err
 			}
 			defer func() { _ = conn.Close(ctx) }()
-			require.NoError(t, runTestQuery(ctx, conn))
-			return true
-		}, 30*time.Second, 100*time.Millisecond)
+			return runTestQuery(ctx, conn)
+		})
 	})
 }
 

--- a/pkg/ccl/sqlproxyccl/server.go
+++ b/pkg/ccl/sqlproxyccl/server.go
@@ -181,7 +181,7 @@ func (s *Server) ServeHTTP(ctx context.Context, ln net.Listener) error {
 
 	srv := http.Server{Handler: s.mux}
 
-	go func() {
+	err := s.Stopper.RunAsyncTask(ctx, "sqlproxy-http-cleanup", func(ctx context.Context) {
 		<-ctx.Done()
 
 		// Wait up to 15 seconds for the HTTP server to shut itself
@@ -196,11 +196,13 @@ func (s *Server) ServeHTTP(ctx context.Context, ln net.Listener) error {
 				// Ignore any errors as this routine will only be called
 				// when the server is shutting down.
 				_ = srv.Shutdown(shutdownCtx)
-
 				return nil
 			},
 		)
-	}()
+	})
+	if err != nil {
+		return err
+	}
 
 	if err := srv.Serve(ln); err != nil && !errors.Is(err, http.ErrServerClosed) {
 		return err

